lncli: `base_fee_msat` and `fee_rate_ppm` for `openchannel`

hieblmi opened the pull request:
## Motivation

Node operators often see their newly confirmed channels' local balance leaking to the remote side especially when connecting to capacity sinks while having configured low default fee rates in `lnd.conf`, e.g. 

- https://twitter.com/bitcoinftm/status/1542083458796748800. 
- https://twitter.com/callebtc/status/1556985524245729283

The issue is also mentioned here https://github.com/lightningnetwork/lnd/pull/6610.

This PR introduces two parameters `base_fee_msat` and `fee_rate` for `lncli openchannel` in order to default to the so specified fees in the channel announcement phase. This way immediate channel leakage can be avoided.

## Implementation

1. Four fields are introduced to `OpenChannelRequest`, baseFee, feeRate, useBaseFee and useFeeRate. The use* fields are added to distinguish between the scenarios where the user provides fee values base_fee 0 / fee_rate 0 on `lncli openchannel` and where the user omits one of the fee values and when the default config parameters for fees should be used in the channel announcement.
2. The `rpcserver.go` parses the fields into `initFundingMsg`
3. Within `funding/manager.go` these new fields are added to a channel reservation context.
4. When the FundingSigned stage is  reached the reservation context is deleted so before that we store the fields per channel id in a new database bucket. Once the channel announcement stage is reached we read the channel fees from the database into the channel announcement structure and delete the fees from the database.
5. In `newChanAnnouncement` we check if we find a fee entry in the database for a completed channel id. If so we apply the fees provided and delete the entry from the bucket.

## Steps to Test
1. See attached `testOpenChannelUpdateFeePolicy` and `funding/manager_test.go`
2. Tested differrent scenarios on testnet while having default fees configured in `lnd.conf`.
```
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40000
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40001 --base_fee_msat 9999
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40002 --fee_rate_ppm 9999
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40003 --base_fee_msat 9999 --fee_rate_ppm 9999
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40004 --base_fee_msat 0 --fee_rate_ppm 9999
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40004 --base_fee_msat 9999 --fee_rate_ppm 0
lnclit openchannel --node_key 030425d8babe3ab6dfc065e69dd8b10ce6738c86ea7d634324c913e21620fa5eaf --local_amt 40004 --base_fee_msat 0 --fee_rate_ppm 0
```

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

hieblmi commented:
@positiveblue @yyforyongyu - the added itest is failing and I believe it is due to the .proto file. Do I need to checkin the generated file after `make rpc` was executed?

hieblmi commented:
Tagging @positiveblue to check the general implementation and the mutex question. I am not sure why the `lightning.pb.go` introduced so many changes. I built the protos with `make rpc`.

hieblmi commented on funding/manager.go:
```diff
@@ -586,6 +615,7 @@ func NewFundingManager(cfg Config) (*Manager, error) {
 		chanIDKey:                   cfg.TempChanIDSeed,
 		activeReservations:          make(map[serializedPubKey]pendingChannels),
 		signedReservations:          make(map[lnwire.ChannelID][32]byte),
+		channelFeeMap:               make(map[lnwire.ChannelID]*htlcswitch.ForwardingPolicy),
```
Line too long, have to wrap this.

hieblmi commented on rpcserver.go:
```diff
@@ -1903,14 +1903,17 @@ func (r *rpcServer) parseOpenChannelReq(in *lnrpc.OpenChannelRequest,
 	// satoshis) does not exceed the amount the local party has requested
 	// for funding.
 	//
-	// TODO(roasbeef): incorporate base fee?
```
@Roasbeef that PR would cover your TODO here I guess?

hieblmi commented on funding/manager.go:
```diff
@@ -3488,6 +3524,19 @@ func (f *Manager) newChanAnnouncement(localPubKey,
+		baseFee = uint32(forwardingPolicy.BaseFee)
+		feeRate = uint32(forwardingPolicy.FeeRate)
+		// Mutex?
+		delete(f.channelFeeMap, chanID)
```
@positiveblue do we need a mutex to read/write `f.channelFeeMap`?

positiveblue commented on lntest/harness_net.go:
```diff
@@ -240,6 +240,7 @@ out:
 
 			if aliceResp.ConfirmedBalance == expectedBalance &&
 				bobResp.ConfirmedBalance == expectedBalance {
+
```
thx

positiveblue commented on lntest/harness_net.go:
```diff
@@ -998,6 +999,12 @@ type OpenChannelParams struct {
 	// channel type negotiation.
 	ScidAlias bool
+
+	// Channel fees applied during the channel announcement phase
```
nit: end sentence with `.`

positiveblue commented on funding/manager.go:
```diff
@@ -3488,6 +3524,19 @@ func (f *Manager) newChanAnnouncement(localPubKey,
 	// max_htlc field.
 	msgFlags := lnwire.ChanUpdateOptionMaxHtlc
 
+	// Assume default fee parameters defined in the config
```
nit: end sentence with `.`

positiveblue commented on funding/manager.go:
```diff
@@ -3488,6 +3524,19 @@ func (f *Manager) newChanAnnouncement(localPubKey,
+
+	// Check if we should use the channel fees provided
+	// in the initFundingMsg.
+	if forwardingPolicy, ok := f.channelFeeMap[chanID]; ok {
```
we could make this the default flow and ensure that all pending channels are always in the map. If they did not have custom fees they are added to the map with the default routing police values. However, we need to be sure that the map gets populated during the lnd starting processes

hieblmi replied:
Hmm, what if a new entry point to this flow is added but the `f.channelFeeMap` is not initialized. How does the implementer of the new flow know that they need to add to the map? Since we can't predict who is calling `newChanAnnouncement` and how they are calling it in the future I would keep the default fees here. Unless you insist :-).

positiveblue replied:
I would be consistent. As a user, if at some point I opened a channel with custom fees and I see it propagated with the default ones I won't be happy with "well, it is just that it followed another flow and the announcer did not have the data with your custom fees".

I think it is better if we return an error when trying to announce a channel if it is not present in `channelFeeMap`.

hieblmi commented on funding/manager.go:
```diff
@@ -517,6 +539,13 @@ type Manager struct {
+	// When the reservation context is deleted in the FundingSigned
+	// handler then we map the permanent channel id here to the
+	// fees that will be announced in a new channel announcement.
+	channelFeeMap map[lnwire.ChannelID]*htlcswitch.ForwardingPolicy
```
As I am unfamiliar with storing fields in buckets in the database I was seeking inspiration from here: https://github.com/lightningnetwork/lnd/blob/d9f4b36cb9009d9e52b952095b612b0edcc3e1e9/funding/manager.go#L4221

Should I follow the example of how the `channelOpeningState` is saved in the DB? Or is there a better example in the code you can point me to on how to store the `channelFeeMap`?

And can we then load the `channelFeeMap` when the funding manager starts? https://github.com/lightningnetwork/lnd/blob/d9f4b36cb9009d9e52b952095b612b0edcc3e1e9/funding/manager.go#L609

Thanks for your help.

hieblmi replied:
Also tagging @guggero here for above question.

guggero replied:
We generally have three options for adding new fields to existing DB objects (with `bbolt` at least).
1. Run a DB migration and add the new field to each existing element in the DB (overkill for this use case IMO because not a lot of users will have pending channels at the time of the `lnd` upgrade and a migration also creates a downgrade barrier which is kind of annoying this early in the 0.16 release schedule).
2. Use TLVs which allows us to add new fields very easily. But since this component doesn't use TLVs yet, switching to TLVs would require a DB migration as well, which we want to avoid.
3. Store new/optional data in a separate key or bucket (depending on the existing structure).

So tl;dr: I agree that we probably should put this field into its own key in `SaveChannelOpeningState` (`channeldb/db.go`).
And while we're at it: I think this property should be stored with each pending channel and not be kept in a global map in the manager. At least I think it makes things easier to understand if the channel knows what fees it should set.
But I haven't taken a closer look at the rest of the PR, so maybe it's not practical to do it in this case? Just drive-by commenting here...

hieblmi replied:
Thanks @guggero, this is helpful info I can ponder on.

Re your suggestion about pending channels - this PR proposes to store the fees for pending channels in their respective reservation context. But once the channel is confirmed the reservation context is deleted. Then we need the global map to map the completed channel id to the fees which is then read in the `newChanAnnouncement` - at least as suggested by my PR so far.

hieblmi replied:
Thanks @guggero, this is helpful info I can ponder on.
> So tl;dr: I agree that we probably should put this field into its own key in `SaveChannelOpeningState` (`channeldb/db.go`). 

Do you suggest that we probably should create an extra bucket for channel fees per channel_id next to`channelOpeningStateBucket = []byte("channelOpeningState")` and update/delete it with a set of new methods, like `SaveChannelFees` etc.?

>And while we're at it: I think this property should be stored with each pending channel and not be kept in a global map in the manager. At least I think it makes things easier to understand if the channel knows what fees it should set.

This PR proposes to store the fees for pending channels in their respective reservation context. But once the channel is in the 'Funding Signed' stage the reservation context is deleted. Then we need the global map to map the completed channel id to the fees which is then read in the `newChanAnnouncement`. So I guess we could replace the global map by storing the channel fee information in the database before the reservation context is deleted?

hieblmi replied:
@guggero, @positiveblue , I've added a new channel fee bucket to the database where fees are persisted and retrieved during the funding process(https://github.com/lightningnetwork/lnd/pull/6753/files#diff-887e88ea126293e649522fa992766c53d32790d424fca8aca4404ad1dc17751d).

I've accompanied my changes with test cases aiming to test fee settings and update and deletion of the db bucket.

hieblmi commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -104,6 +104,20 @@ var openChannelCommand = cli.Command{
+				"of payment size",
+		},
+		cli.Int64Flag{
+			Name: "fee_rate",
```
Agree, will change it to `fee_rate_ppm`.

SachinMeier reviewed (commented):
This is a great change! Excited to see it merged.

hieblmi commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -104,6 +104,19 @@ var openChannelCommand = cli.Command{
 			Name:  "local_amt",
 			Usage: "the number of satoshis the wallet should commit to the channel",
 		},
+		cli.Int64Flag{
+			Name: "base_fee_msat",
+			Usage: "the base fee in milli-satoshis that will " +
+				"be charged for each forwarded HTLC, regardless " +
+				"of payment size",
+		},
+		cli.Int64Flag{
```
Thanks for that hint, I will fix it.

hieblmi commented on funding/manager.go:
```diff
@@ -197,6 +200,25 @@ type InitFundingMsg struct {
 	// LocalFundingAmt is the size of the channel.
 	LocalFundingAmt btcutil.Amount
 
+	// The base fee charged regardless of the number of milli-satoshis sent.
```
Appreciate the review, will fix your findings.

hieblmi commented on docs/release-notes/release-notes-0.15.1.md:
```diff
@@ -30,6 +30,11 @@
   user to request that specific directional node pairs be ignored during the 
   route search.
 
+* [Add `base_fee_msat` and `fee_rate` flags to
```
Moved to 0.16.0. - remove this block

hieblmi commented on funding/manager.go:
```diff
@@ -2158,6 +2180,14 @@ func (f *Manager) handleFundingCreated(peer lnpeer.Peer,
 		return
 	}
 
+	// With a permanent channel id established we can save it respective
```
typo

hieblmi commented on funding/manager.go:
```diff
@@ -3616,6 +3678,13 @@ func (f *Manager) announceChannel(localIDKey, remoteIDKey *btcec.PublicKey,
 	}
 
+	// After the fee parameters have been stored to the announcement ann
+	// we can delete them in the database.
```
grammar

positiveblue commented on rpcserver.go:
```diff
@@ -1910,12 +1910,26 @@ func (r *rpcServer) parseOpenChannelReq(in *lnrpc.OpenChannelRequest,
+	if in.UseBaseFee {
+		channelBaseFee = &in.BaseFee
+	} else {
+		rpcsLog.Infof("hiebmi: rpcserver, no basefee specified from client")
```
I think we can delete these.

positiveblue commented on funding/manager_test.go:
```diff
@@ -1041,11 +1041,12 @@ func assertAddedToRouterGraph(t *testing.T, alice, bob *testNode,
 func assertChannelAnnouncements(t *testing.T, alice, bob *testNode,
 	capacity btcutil.Amount, customMinHtlc []lnwire.MilliSatoshi,
-	customMaxHtlc []lnwire.MilliSatoshi) {
+	customMaxHtlc []lnwire.MilliSatoshi, baseFees []lnwire.MilliSatoshi,
```
nit: maybe we could add here something like

```
type channelAnnouncementCheck func(msg *lnwire.Message) error

func assertChannelAnnouncements(t *testing.T, alice, bob *testNode,
capacity btcutil.Amount, customMinHtlc []lnwire.MilliSatoshi,
customMaxHtlc []lnwire.MilliSatoshi, additionalChecks ...ChannelAnnouncementCheck) {
    // ...
    for _, check := range additionalChecks {
        check(msg)
    }
}
```

So we don't have to pass most of the checks when they are not needed. It can be added later so no worries.

hieblmi replied:
Yes it makes sense to condense the parameter list of `assertChannelAnnouncements` the way you describe. We could   probably take `customMinHtlc`, `customMaxHtlc`, `baseFees` and `feeRates` and put them into a `htlc.ForwardingPolicy` for each Alice and Bob and pass it to the `additionalChecks` function where necessary. 

If you agree then I would keep the test as is in this PR and tackle a refactoring of it in a follow-up PR.

positiveblue commented on lntest/itest/lnd_open_channel_test.go:
```diff
@@ -249,6 +250,150 @@ func testOpenChannelAfterReorg(net *lntest.NetworkHarness, t *harnessTest) {
+//		ChannelUpdate --> defaultBaseFee, defaultFeeRate
+// 2.) only baseFee provided to OpenChannelRequest
+//		ChannelUpdate --> provided baseFee, defaultFeeRate
+// 3.) no params provided to OpenChannelRequest
```
This comment needs to be fixed

hieblmi replied:
Good catch, thanks.

positiveblue commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -104,6 +104,19 @@ var openChannelCommand = cli.Command{
 			Name:  "local_amt",
 			Usage: "the number of satoshis the wallet should commit to the channel",
 		},
+		cli.Int64Flag{
```
can these ones be `cli.Uint64Flag`?

positiveblue commented on channeldb/db.go:
```diff
@@ -245,6 +245,11 @@ var (
+	// channelFeeBucket is the database bucket used to store the
+	// channel fees for each permanent channel that is currently in the process
+	// of being opened.
+	channelFeeBucket = []byte("channelFees")
```
Q: will these bucket be reused for other custom forward policies like minhtlc? It can be added in another PR (if needed) but I wonder if it is better to call the bucket something else like "channelInitialForwardPolicy"

hieblmi replied:
Yes, imo ideally all fields from `htlc.ForwardingPolicy` should be able to be specified from `lncli` and be announced and saved like the channel fees here. This enhancement would probably go well in another PR with the test refactoring above in `funding/manager_test.go`. I'll rename the bucket to `initialChannelForwardingPolicy`.

positiveblue reviewed (commented):
I am really excited about this! Congratulations @hieblmi this is looking pretty good :tada: 


I left some small comments. I will do a final pass after all the doubts are cleared.

hieblmi commented:
@positiveblue I've addressed the issues you raised in the latest commit.

positiveblue reviewed (changes requested):
Great work @hieblmi I think we are almost done here, really solid contribution :fire: 

Small fixes, nits before green light:
- We should reorg the commits so there is the rpc definition (and compilation) before using it in the cmd.
- I see a change on the release notes for v0.15.1, if this is for v0.16.0 that other change should go in a different PR or it will not land v0.15.1, even if it is just sorting the names.
- I would change the naming of some functions from `xxxFees` to `xxxInitialFwdPolicy`
- We need to keep in mind what happens in case:
    - I open a channel in version <v0.15.1
    - I update to v0.15.1 before that channel is confirmed
    - The channel gets confirmed after restarting the deamon in v0.15.1

hieblmi commented on funding/manager.go:
```diff
@@ -3504,8 +3566,8 @@ func (f *Manager) newChanAnnouncement(localPubKey,
+		FeeRate: feeRate,
 	}
 
 	if ourPolicy != nil {
```
Yes I like your solution. It's more readable than the current version - `ourPolicy` takes still precedence over `storedFwdPolicy`... I will apply these changes.

hieblmi commented on lntest/itest/lnd_open_channel_test.go:
```diff
@@ -249,6 +250,149 @@ func testOpenChannelAfterReorg(net *lntest.NetworkHarness, t *harnessTest) {
+//		ChannelUpdate --> defaultBaseFee, provided FeeRate
+// 4.) baseFee and feeRate provided to OpenChannelRequest
+//		ChannelUpdate --> provided baseFee, privided feeRate
+func testOpenChannelUpdateFeePolicy(net *lntest.NetworkHarness,
```
That is a very good catch. I think in that case we would land in the `default:` branch of our suggested `switch` statement above since we were not storing any fees(provided or default) in the db and `ourPolicy` might not be populated. This edge case would obviously be solved by setting the default fees in `newChanAnnouncement` but that would defeat the purpose of this change.

positiveblue replied:
in this case we have two options:
- A) DB migration, setting the defaults for the channels that are currently pending
- B) if we do not know about that channel, use the defaults


I think we want to catch the cases in B) as "something went wrong here" but I would avoid A) because a migration does not look worth it here.

I would go with B, add a log so we can track what is going on in case someone finds something broken and switch to cancel if we do not have `storedFwdPolicy` in the future.

hieblmi replied:
If we go with B) it would mean that we wouldn't return an error in `newChanAnnouncement` if both `ourPolicy` and `storedFwdingPolicy` are `nil`, but that we only log this scenario.

hieblmi commented on funding/manager.go:
```diff
@@ -4265,6 +4345,60 @@ func copyPubKey(pub *btcec.PublicKey) *btcec.PublicKey {
 	return btcec.NewPublicKey(&tmp.X, &tmp.Y)
 }
 
+// saveInitialFwdingPolicy saves the channelOpeningState for the provided
```
fix comment

lightninglabs-deploy commented:
@positiveblue: review reminder
@hieblmi, remember to re-request review from reviewers when ready

positiveblue commented on lnrpc/lightning.proto:
```diff
@@ -2189,6 +2189,33 @@ message OpenChannelRequest {
+    The effective fee rate in milli-satoshis. The precision of this value
+    goes up to 6 decimal places, so 1e-6.
+    */
+    uint64 fee_rate = 22;
```
Not sure if what we want here is the amount in milli-satoshis or the same granularity that we have in the config file (ppm directly)

https://github.com/lightningnetwork/lnd/blob/master/sample-lnd.conf#L524

positiveblue commented on funding/manager.go:
```diff
@@ -3983,6 +4055,22 @@ func (f *Manager) handleInitFundingMsg(msg *InitFundingMsg) {
+	}
+
+	if feeRate != nil {
+		forwardingPolicy.FeeRate = lnwire.MilliSatoshi(*feeRate)
```
if we change this to PPM we need to be sure that we change it to MilliSatoshis :+1:

hieblmi replied:
Sorry I am not sure I understand this comment.

`forwardingPolicy.FeeRate` is a `uint64` currently denoting the fee rate in milli satoshis. Since the `forwardingPolicy` doesn't have a fee field expressed in ppm this one is used to work around it. Are you suggesting that we add a new field for ppm to `ForwardingPolicy`?

positiveblue replied:
Sorry that was confusing. I just meant that we should avoid phrasing this in the api (proto) as "The effective fee rate in milli-satoshis."

 This is a uint64 that means parts per million (independently of how we treat it inside lnd). We use `lnwire.MilliSatoshi` to represent it because we want to multiply it by the amount (that is `lnwire.MiliSatoshi`), but if you say `feeRate=1` you would get 1 btc if you forward one milion btc, the same way you would get 1 sat if you foward 1M sats and so on.

positiveblue commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -104,6 +104,19 @@ var openChannelCommand = cli.Command{
+			Usage: "the fee rate that will be charged " +
+				"proportionally based on the value of each " +
+				"forwarded HTLC, the lowest possible rate is 0 " +
+				"with a granularity of 0.000001 (millionths).",
```
I would change this to make it explicit that this will be the rate per million. If not it looks a bit weird that you have a granularity of `0.000001` in a uint64.

positiveblue reviewed (commented):
Looking good :+1: I left what look like the last comments before my :heavy_check_mark:

positiveblue commented on funding/manager.go:
```diff
@@ -2174,6 +2198,14 @@ func (f *Manager) handleFundingCreated(peer lnpeer.Peer,
 
+	// With a permanent channel id established we can save the respective
+	// forwarding policy in the database. In the channel announcement phase
+	// this forwarding policiy is retrieved and applied.
```
s/policiy/policy

positiveblue commented on funding/manager.go:
```diff
@@ -3571,9 +3612,33 @@ func (f *Manager) newChanAnnouncement(localPubKey,
+		chanUpdateAnn.BaseFee = uint32(f.cfg.DefaultRoutingPolicy.BaseFee)
+		chanUpdateAnn.FeeRate = uint32(f.cfg.DefaultRoutingPolicy.FeeRate)
+
+		// Decide between storing the defaults and using them here or
```
`TODO: Decide...`

positiveblue commented on funding/manager.go:
```diff
@@ -3670,6 +3735,13 @@ func (f *Manager) announceChannel(localIDKey, remoteIDKey *btcec.PublicKey,
+	// we can delete them from the database.
+	err = f.deleteInitialFwdingPolicy(chanID)
+	if err != nil {
+		log.Infof("Could not delete channel fees for chanId %x\n.", chanID)
```
`log.Infof` already includes an end of line `\n`

positiveblue commented on funding/manager.go:
```diff
@@ -4265,6 +4356,60 @@ func copyPubKey(pub *btcec.PublicKey) *btcec.PublicKey {
+}
+
+// getInitialFwdingPolicy fetches channel fees for a given channel id from
+// the during the channel announcement.
```
nit: incomplete sentence?

positiveblue commented on lntest/itest/lnd_open_channel_test.go:
```diff
@@ -249,6 +250,149 @@ func testOpenChannelAfterReorg(net *lntest.NetworkHarness, t *harnessTest) {
+		nodes := []*lntest.HarnessNode{net.Alice, net.Bob}
+
+		// Alice and Bob should see each other's ChannelUpdates, advertising the
+		// default routing policies.
```
s/default routing policies/preferred routing policies?

positiveblue reviewed (approved):
LGTM :rocket: 

I added a couple of extra nits, but I think this is ready :+1: great work @hieblmi this PR will make many users happy :pray:

sputn1ck commented on funding/manager.go:
```diff
@@ -3670,6 +3735,13 @@ func (f *Manager) announceChannel(localIDKey, remoteIDKey *btcec.PublicKey,
 
+	// After the fee parameters have been stored in the announcement
+	// we can delete them from the database.
+	err = f.deleteInitialFwdingPolicy(chanID)
```
In the pr review club we noticed that this is only called when the channel is announced. If we create a private channel with custom feerates, this will never be called and the forwarding policy would never be deleted.

sputn1ck reviewed (commented):
Very useful PR, during the review club we noticed that the stored forwarding policies only get deleted when the channel is announced, meaning we never delete it for private channels with custom feerates.

hieblmi commented:
@sputn1ck @positiveblue 
I've added the deletion of the forwarding policy in case of a private channel funding flow:
https://github.com/lightningnetwork/lnd/blob/a948ba91db697e14c9b7b26e26fee969c7202aa9/funding/manager.go#L3127

I've tested this change in two ways. Firstly opened a private channel on testnet:
1. Open Private Channel
```
lnclit openchannel --node_key 02340ceae1c0e1779577a947ff1b09df9962e20dccfad31772d437501c31eca28f --local_amt 20004 --base_fee_msat 3333 --fee_rate_ppm 3333 --private
{
	"funding_txid": "b2128cd17b867a34f3b5f394a06aa602c6ee3e6943da8534887bf6453cfffa9e"
}
```
2. Check the bolt `channel.db` with [https://github.com/guggero/boltbrowser](boltbrowser) and see an entry for the forwarding policy:
<img width="744" alt="image" src="https://user-images.githubusercontent.com/18631614/189493022-75f5081b-0b20-4cc9-8e07-7bebe07fe8b3.png">
3. Wait for 6 confirmations and check the channel.db again.
<img width="744" alt="image" src="https://user-images.githubusercontent.com/18631614/189493054-2e85ee7d-bc49-4ed6-a661-a24f794d2a69.png">

Secondly I added additional checks in `manager_test.go` to make sure that there are is no forwarding policies saved in the db at the end of each test.
https://github.com/lightningnetwork/lnd/blob/188ad1cad9a8116896d4d96e00795280875718e9/funding/manager_test.go#L1293

positiveblue commented on funding/manager.go:
```diff
@@ -3080,6 +3120,14 @@ func (f *Manager) annAfterSixConfs(completeChan *channeldb.OpenChannel,
 				"to peer %x: %v", pubKey, err)
 		}
+
+		// For private channels we do not announce the channel policy
```
good catch @sputn1ck :+1:

positiveblue commented on channeldb/db.go:
```diff
@@ -1290,6 +1295,61 @@ func (c *ChannelStateDB) AbandonChannel(chanPoint *wire.OutPoint,
+		bucket := tx.ReadBucket(initialChannelFwdingPolicyBucket)
+		if bucket == nil {
+			// If the bucket does not exist, it means we never added
+			//  a channel fees to the db, so return ErrChannelNotFound.
```
nit: extra space

positiveblue commented on funding/manager_test.go:
```diff
@@ -2505,6 +2607,10 @@ func TestFundingManagerPrivateChannel(t *testing.T) {
 	assertNoChannelState(t, alice, bob, fundingOutPoint)
+
+	// The forwarding policy for the channel announcement should
+	// have been deleted from the database, as the channel is announced.
```
nit : rm `as the channel is announced` as it will never happen for private channels

positiveblue commented on funding/manager_test.go:
```diff
@@ -2640,6 +2746,10 @@ func TestFundingManagerPrivateRestart(t *testing.T) {
 	assertNoChannelState(t, alice, bob, fundingOutPoint)
+
+	// The forwarding policy for the channel announcement should
+	// have been deleted from the database, as the channel is announced.
```
nit: same here

positiveblue reviewed (approved):
LGTM :rocket: let's wait for a second review from @sputn1ck

hieblmi commented:
Addressed nits in latest change set.

sputn1ck reviewed (approved):
tACK! LGTM 🚀

guggero commented on lnrpc/lightning.proto:
```diff
@@ -2190,6 +2190,33 @@ message OpenChannelRequest {
+    uint64 fee_rate = 22;
+
+    /*
+    If UseBaseFee is true the open channel announcement will update the
```
We normally use the notation of the proto variables and not the resulting Golang names. So we should use `use_base_fee` and `base_fee` in the comments instead of `UseBaseFee`.

guggero commented on lnrpc/lightning.proto:
```diff
@@ -2190,6 +2190,33 @@ message OpenChannelRequest {
     */
     bool scid_alias = 20;
+
+    /*
```
Commit message prefix should be `lnrpc`.

guggero commented on channeldb/db.go:
```diff
@@ -276,6 +276,11 @@ var (
 	// of being opened.
 	channelOpeningStateBucket = []byte("channelOpeningState")
+
+	// initialChannelForwardingPolicy is the database bucket used to store the
```
Commit message prefix should be `funding+channeldb:` then the text itself can just be `handle and persist channel fees` for example.

guggero commented on funding/manager.go:
```diff
@@ -1590,6 +1602,14 @@ func (f *Manager) handleFundingOpen(peer lnpeer.Peer,
 
+	// If we are handling a FundingOpen request then we need to
+	// specify the default channel fees since they are not provided
+	// by the responder interactively
```
nit: comments should always end with full stops.

guggero commented on funding/manager.go:
```diff
@@ -3080,6 +3120,14 @@ func (f *Manager) annAfterSixConfs(completeChan *channeldb.OpenChannel,
+		// database.
+		err = f.deleteInitialFwdingPolicy(chanID)
+		if err != nil {
+			log.Infof("Could not delete channel fees for chanId %x.", chanID)
```
nit: line length.

guggero commented on funding/manager.go:
```diff
@@ -3571,9 +3620,33 @@ func (f *Manager) newChanAnnouncement(localPubKey,
+		chanUpdateAnn.FeeRate = uint32(storedFwdingPolicy.FeeRate)
+
+		// TODO: Decide between storing the defaults and using them here or
+		// simply put the defaults that where used until now. Right now
```
nit: s/where/were.

guggero reviewed (commented):
Thanks for the PR. Latest state looks pretty good!
Found a few nits though that we should try to get rid of.

hieblmi commented:
@guggero addressed the nits in latest change set.

guggero commented:
Thanks for addressing the nits in all commits, not just the ones I pointed out. Though the formatting is off, please see https://github.com/lightningnetwork/lnd/blob/master/docs/code_formatting_rules.md (part of the checklist in the PR template).

hieblmi commented:
@guggero I think I caught all formatting issues and will keep them in mind.

guggero commented on funding/manager.go:
```diff
@@ -2110,6 +2131,9 @@ func (f *Manager) handleFundingCreated(peer lnpeer.Peer,
 		return
 	}
 
+	// Get forwarding policy before deleting the reservation context
```
nit: missing full stop.

guggero commented on lnrpc/lightning.proto:
```diff
@@ -2190,6 +2190,33 @@ message OpenChannelRequest {
+
+    /*
+    If use_fee_rate is true the open channel announcement will update the
+    channel fee rate with the value specified in FeeRate. In the case of
```
nit: s/FeeRate/fee_rate

guggero reviewed (commented):
Last set of nits, then we're good to go :rocket:

hieblmi commented:
@guggero addressed latest nits.

guggero commented on lntest/harness_net.go:
```diff
@@ -998,6 +999,12 @@ type OpenChannelParams struct {
 	// channel type negotiation.
 	ScidAlias bool
+
+	// Channel fees applied during the channel announcement phase.
```
I thought I already added this comment in my last review, but it looks like I forgot. But those struct fields all need an individual Godoc comment.

guggero commented on funding/manager.go:
```diff
@@ -3576,6 +3640,41 @@ func (f *Manager) newChanAnnouncement(localPubKey,
+		chanUpdateAnn.BaseFee = uint32(storedFwdingPolicy.BaseFee)
+		chanUpdateAnn.FeeRate = uint32(storedFwdingPolicy.FeeRate)
+
+		// TODO: Decide between storing the defaults and using them here
```
The question about the TODO is still open. Is this something that actually needs to be decided or can we remove the TODO instead?
Also, the lines following after the comment are now duplicated between this case and the default case. If we set the values when we initialize `chanUpdateAnn` and only overwrite what we can actually configure this `switch` statement becomes easier to read and understand IMO.

guggero reviewed (approved):
LGTM :tada:

hieblmi commented:
Unreal 🍾🎉. Thanks yalls for reviewing and teaching this flow!
